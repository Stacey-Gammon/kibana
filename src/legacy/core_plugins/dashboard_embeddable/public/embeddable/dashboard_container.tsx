/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React from 'react';
import ReactDOM from 'react-dom';
import uuid from 'uuid';

import { I18nProvider } from '@kbn/i18n/react';
import { IndexPattern } from 'ui/index_patterns';

import {
  Container,
  ContainerInput,
  ContainerOutput,
  Embeddable,
  EmbeddableFactoryRegistry,
  EmbeddableInput,
  Filter,
  Query,
  RefreshConfig,
  ViewMode,
  EmbeddableOutput,
} from 'plugins/embeddable_api/index';
import { TimeRange } from 'ui/timefilter/time_history';

import { EmbeddableInputMissingFromContainer } from 'plugins/embeddable_api/containers/container';
import { DASHBOARD_CONTAINER_TYPE } from './dashboard_container_factory';
import { createPanelState } from './panel';
import { DashboardPanelState } from './types';
import { DashboardViewport } from './viewport/dashboard_viewport';

export interface DashboardContainerInput extends ContainerInput {
  viewMode: ViewMode;
  filters: Filter[];
  query: Query;
  timeRange: TimeRange;
  refreshConfig?: RefreshConfig;
  expandedPanelId?: string;
  useMargins: boolean;
  title: string;
  description?: string;
  isFullScreenMode: boolean;
  panels: { [panelId: string]: DashboardPanelState };

  // Used to force a refresh of embeddables even if there were no other input state
  // changes.
  lastReloadRequestTime?: number;
}

export interface DashboardEmbeddableInput extends EmbeddableInput {
  // customization: any;
  filters: Filter[];
  isPanelExpanded: boolean;
  query: Query;
  timeRange: TimeRange;
  refreshConfig?: RefreshConfig;
  viewMode: ViewMode;
  savedObjectId?: string;
  hidePanelTitles?: boolean;
}

export interface DashboardEmbeddableOutput extends EmbeddableOutput {
  indexPatterns?: IndexPattern[];
}

export type DashboardEmbeddable = Embeddable<DashboardEmbeddableInput, DashboardEmbeddableOutput>;

export class DashboardContainer extends Container<
  DashboardEmbeddableInput,
  DashboardEmbeddableOutput,
  DashboardContainerInput
> {
  constructor(
    initialInput: DashboardContainerInput,
    embeddableFactories: EmbeddableFactoryRegistry
  ) {
    super(DASHBOARD_CONTAINER_TYPE, initialInput, { embeddableLoaded: {} }, embeddableFactories);
  }

  public createNewPanelState<EEI extends EmbeddableInput = EmbeddableInput>({
    type,
    initialInput,
  }: {
    type: string;
    initialInput: EmbeddableInputMissingFromContainer<EEI, DashboardEmbeddableInput> & {
      id?: string;
    };
  }): DashboardPanelState<EmbeddableInputMissingFromContainer<EEI, DashboardEmbeddableInput>> {
    const embeddableId = initialInput.id || uuid.v4();
    return createPanelState<EmbeddableInputMissingFromContainer<EEI, DashboardEmbeddableInput>>(
      { ...initialInput, id: embeddableId },
      type,
      Object.values(this.input.panels)
    );
  }

  public onToggleExpandPanel(id: string) {
    const newValue = this.input.expandedPanelId ? undefined : id;
    this.updateInput({
      expandedPanelId: newValue,
    });
  }

  public onPanelsUpdated = (panels: { [panelId: string]: DashboardPanelState }) => {
    this.updateInput({
      panels: {
        ...panels,
      },
    });
  };

  public onExitFullScreenMode = () => {
    this.updateInput({
      isFullScreenMode: false,
    });
  };

  public render(dom: React.ReactNode) {
    ReactDOM.render(
      // @ts-ignore
      <I18nProvider>
        <DashboardViewport embeddableFactories={this.embeddableFactories} container={this} />
      </I18nProvider>,
      dom
    );
  }

  public getPanelIndexPatterns() {
    const indexPatterns: IndexPattern[] = [];
    Object.values(this.embeddables).forEach(embeddable => {
      const embeddableIndexPatterns = embeddable.getOutput().indexPatterns;
      if (embeddableIndexPatterns) {
        indexPatterns.push(...embeddableIndexPatterns);
      }
    });
    return indexPatterns;
  }

  public getInputForEmbeddableFromContainer(id: string): DashboardEmbeddableInput {
    const panelState = this.getPanelState(id);
    const isPanelExpanded = this.input.expandedPanelId === panelState.embeddableId;
    const { viewMode, refreshConfig, timeRange, query, hidePanelTitles, filters } = this.input;
    return {
      filters,
      hidePanelTitles,
      isPanelExpanded,
      query,
      timeRange,
      refreshConfig,
      viewMode,
      id: panelState.embeddableId,
      customization: panelState.customization,
    };
  }
}
