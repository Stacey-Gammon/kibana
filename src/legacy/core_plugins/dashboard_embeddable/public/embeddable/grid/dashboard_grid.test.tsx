/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React from 'react';
import { shallowWithIntl } from 'test_utils/enzyme_helpers';
// @ts-ignore
import sizeMe from 'react-sizeme';

import { ViewMode, EmbeddableFactoryRegistry } from 'plugins/embeddable_api/index';

import { DashboardGrid, DashboardGridProps } from './dashboard_grid';
import {
  HelloWorldEmbeddableFactory,
  HELLO_WORLD_EMBEDDABLE,
} from 'plugins/embeddable_api/__test__';
import { DashboardContainer } from '../dashboard_container';
import { getSampleDashboardInput } from 'plugins/dashboard_embeddable/__test__';

jest.mock('ui/chrome', () => ({ getKibanaVersion: () => '6.0.0' }), { virtual: true });

jest.mock(
  'ui/notify',
  () => ({
    toastNotifications: {
      addDanger: () => {},
    },
  }),
  { virtual: true }
);

let dashboardContainer: DashboardContainer | undefined;

function getProps(props?: Partial<DashboardGridProps>): DashboardGridProps {
  const embeddableFactories = new EmbeddableFactoryRegistry();
  embeddableFactories.registerFactory(new HelloWorldEmbeddableFactory());
  dashboardContainer = new DashboardContainer(
    getSampleDashboardInput({
      panels: {
        '1': {
          gridData: { x: 0, y: 0, w: 6, h: 6, i: '1' },
          embeddableId: '1',
          type: HELLO_WORLD_EMBEDDABLE,
          initialInput: { firstName: 'Bob' },
        },
        '2': {
          gridData: { x: 6, y: 6, w: 6, h: 6, i: '2' },
          type: HELLO_WORLD_EMBEDDABLE,
          embeddableId: '2',
          initialInput: { firstName: 'Stacey' },
        },
      },
    }),
    embeddableFactories
  );
  const defaultTestProps: DashboardGridProps = {
    dashboardViewMode: ViewMode.EDIT,
    embeddableFactories,
    onPanelsUpdated: () => {},
    useMargins: true,
    container: dashboardContainer,
    intl: null as any,
  };
  return Object.assign(defaultTestProps, props);
}

beforeAll(() => {
  // sizeme detects the width to be 0 in our test environment. noPlaceholder will mean that the grid contents will
  // get rendered even when width is 0, which will improve our tests.
  sizeMe.noPlaceholders = true;
});

afterAll(() => {
  sizeMe.noPlaceholders = false;
});

test('renders DashboardGrid', () => {
  const component = shallowWithIntl(<DashboardGrid.WrappedComponent {...getProps()} />);
  expect(component).toMatchSnapshot();
  const panelElements = component.find('Connect(InjectIntl(DashboardPanelUi))');
  expect(panelElements.length).toBe(2);
});

test('renders DashboardGrid with no visualizations', () => {
  const props = getProps();
  props.container.updateInput({ panels: {} });
  const component = shallowWithIntl(<DashboardGrid.WrappedComponent {...props} />);
  expect(component).toMatchSnapshot();
});
