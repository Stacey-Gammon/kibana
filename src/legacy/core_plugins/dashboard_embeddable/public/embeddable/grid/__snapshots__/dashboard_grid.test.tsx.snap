// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders DashboardGrid 1`] = `""`;

exports[`renders DashboardGrid with no visualizations 1`] = `""`;
